tiny_dnn/layers/l2_normalization_layer.h

```diff
@@ -0,0 +1,141 @@
+   * @param epsilon         [in] small positive value to avoid zero-division
+   * @param momentum        [in] momentum in the computation of the exponential
+   *average of the mean/stddev of the data
+   * @param phase           [in] specify the current context (train/test)
+   **/
+  l2_normalization_layer(const layer &prev_layer,
+                            float_t epsilon  = 1e-10,
+                            float_t scale = 20)
+    : Base({vector_type::data}, {vector_type::data}),
+      in_channels_(prev_layer.out_shape()[0].depth_),
+      in_spatial_size_(prev_layer.out_shape()[0].area()),
+      eps_(epsilon),
+      scale_(scale) {
+  }
+
+  /**
+   * @param in_spatial_size [in] spatial size (WxH) of the input data
+   * @param in_channels     [in] channels of the input data
+   * @param epsilon         [in] small positive value to avoid zero-division
+   * @param momentum        [in] momentum in the computation of the exponential
+   *average of the mean/stddev of the data
+   * @param phase           [in] specify the current context (train/test)
+   **/
+  l2_normalization_layer(size_t in_spatial_size,
+                            size_t in_channels,
+                            float_t epsilon = 1e-10,
+                            float_t scale = 20)
+    : Base({vector_type::data}, {vector_type::data}),
+      in_channels_(in_channels),
+      in_spatial_size_(in_spatial_size),
+      eps_(epsilon),
+      scale_(scale) {
+  }
+
+  virtual ~l2_normalization_layer() {}
+
+  ///< number of incoming connections for each output unit
+  size_t fan_in_size() const override { return 1; }
+
+  ///< number of outgoing connections for each input unit
+  size_t fan_out_size() const override { return 1; }
+
+  std::vector<index3d<size_t>> in_shape() const override {
+    return {index3d<size_t>(in_spatial_size_, 1, in_channels_)};
+  }
+
+  std::vector<index3d<size_t>> out_shape() const override {
+    return {index3d<size_t>(in_spatial_size_, 1, in_channels_)};
+  }
+
+  void forward_propagation(const std::vector<tensor_t *> &in_data,
+                           std::vector<tensor_t *> &out_data) override {
+    tensor_t &in  = *in_data[0];
+    tensor_t &out = *out_data[0];
+
+    for_i(in_data[0]->size(), [&](size_t i) {
+      const float_t *inptr = &in[i][0];
+      float_t *outptr      = &out[i][0];
+
+      for (size_t j = 0; j < in_spatial_size_; ++j) {
+        float sum_of_square = 0;
+        for (size_t k = 0; k < in_channels_; ++k) {
+          float value = *(inptr + k * in_spatial_size_);
+          sum_of_square += value * value;
+        }
+        if ( sum_of_square < eps_ ) {
+          sum_of_square = eps_;
+        }
+        sum_of_square = sqrt(sum_of_square);
+
+        for (size_t k = 0; k < in_channels_; ++k) {
+          const float_t *inptr_c = inptr + k * in_spatial_size_;
+          float_t *outptr_c = outptr + k * in_spatial_size_;
+          *outptr_c = *inptr_c / sum_of_square * scale_;
```

Comment:
@beru 
Yes. But how to solve problem?
How about

```
if (eps_==0) {
  eps_ = 1e-12;
}
```

Reply:
@hzxie I'm not really sure. Maybe adding this code at constructor might be good.

```cpp
eps_(std::max(epsilon, std::numeric_limits<float_t>::epsilon())),
```

and also changing the code in this way might help since `sqrt` may produce the value smaller than the `eps_`.
```cpp
        sum_of_square = std::max(sqrt(sum_of_square), eps_);
```

Precisely, it's not `sum_of_square` after calling `sqrt`. So you can introduce another local variable.
But I'm not smart enough to think of proper name....

Reply:
I don't think we need to introduce another local variable.
Because `eps_` is used to avoid zero-division, there is little difference between sum_of_square before and after calling sqrt.

Reply:
How about `rss` or `root_sum_squared`? After calling `sqrt`, the meaning of the value changes, so It might help clarify operations.